# Statistical Methods in Pandas - Lab

## Introduction

In this lesson you'll get some hands-on experience using some of the key summary statistics methods in Pandas.

## Objectives
You will be able to:

* Understand and use the df.describe() and df.info() summary statistics methods
* Use built-in Pandas methods for calculating summary statistics (.mean(), .std(), .count(), .sum(), .mean(), .median(), .std(), .var() and .quantile())
* Apply a function to every element in a Series or DataFrame using s.apply() and df.applymap()


## Getting Started

For this lab, we'll be working with a dataset containing information on various lego datasets.  You will find this dataset in the file `lego_sets.csv`.  

In the cell below:

* Import pandas and set the standard alias of `pd`
* Load in the `lego_sets.csv` dataset using the `read_csv()` function
* Display the head of the DataFrame to get a feel for what we'll be working with

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('lego_sets.csv')

In [3]:
df.head(2)

ages  list_price  num_reviews  piece_count  play_star_rating  \
0  6-12       29.99          2.0        277.0               4.0   
1  6-12       19.99          2.0        168.0               4.0   

                                           prod_desc  prod_id  \
0  Catapult into action and take back the eggs fr...  75823.0   
1  Launch a flying attack and rescue the eggs fro...  75822.0   

                                      prod_long_desc review_difficulty  \
0  Use the staircase catapult to launch Red into ...           Average   
1  Pilot Pig has taken off from Bird Island with ...              Easy   

                set_name  star_rating    theme_name  val_star_rating country  
0  Bird Island Egg Heist          4.5  Angry Birds™              4.0      US  
1     Piggy Plane Attack          5.0  Angry Birds™              4.0      US

## Getting DataFrame-Level Statistics

We'll begin by getting some overall summary statistics on the dataset.  There are two ways we'll get this information-- `.info()` and `.describe()`.

### Using `.info()`

The `.info()` method provides us metadata on the DataFrame itself.  This allows to answer questions such as:

* What data type does each column contain?
* How many rows are in my dataset? 
* How many total non-missing values does each column contain?
* How much memory does the DataFrame take up?

In the cell below, call our DataFrame's `.info()` method. 

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 14 columns):
ages                 12261 non-null object
list_price           12261 non-null float64
num_reviews          10641 non-null float64
piece_count          12261 non-null float64
play_star_rating     10486 non-null float64
prod_desc            11884 non-null object
prod_id              12261 non-null float64
prod_long_desc       12261 non-null object
review_difficulty    10206 non-null object
set_name             12261 non-null object
star_rating          10641 non-null float64
theme_name           12258 non-null object
val_star_rating      10466 non-null float64
country              12261 non-null object
dtypes: float64(7), object(7)
memory usage: 1.3+ MB


#### Interpreting the Results

Read the output above, and then answer the following questions:

How many total rows are in this DataFrame?  How many columns contain numeric data? How many contain categorical data?  Identify at least 3 columns that contain missing values. 

Write your answer below this line:
________________________________________________________________________________________________________________________________

12261 rows
7 numeric cols
7 string (categorical)
num_reviews, play_star_rating, val_star_rating are missing data.

## Using `.describe()`

Whereas `.info()` provides statistics about the DataFrame itself, `.describe()` returns output containing basic summary statistics about the data contained with the DataFrame.  

In the cell below, call the DataFrame's `.describe()` method. 

In [6]:
df.describe()

list_price   num_reviews   piece_count  play_star_rating  \
count  12261.000000  10641.000000  12261.000000      10486.000000   
mean      65.141998     16.826238    493.405921          4.337641   
std       91.980429     36.368984    825.364580          0.652051   
min        2.272400      1.000000      1.000000          1.000000   
25%       19.990000      2.000000     97.000000          4.000000   
50%       36.587800      6.000000    216.000000          4.500000   
75%       70.192200     13.000000    544.000000          4.800000   
max     1104.870000    367.000000   7541.000000          5.000000   

            prod_id   star_rating  val_star_rating  
count  1.226100e+04  10641.000000     10466.000000  
mean   5.983675e+04      4.514134         4.228960  
std    1.638115e+05      0.518865         0.660282  
min    6.300000e+02      1.800000         1.000000  
25%    2.103400e+04      4.300000         4.000000  
50%    4.206900e+04      4.700000         4.300000  
75%    7.092200e+04      5.000000         4.700000  
max    2.000431e+06      5.000000         5.000000

825 std piece count.
7541 max lego count.
1 min lego count.
4.3 median star val count

#### Interpreting the Results

The output contains descriptive statistics corresponding to the columns.  Use these to answer the following questions:

How much is the standard deviation for piece count?  How many pieces are in the largest lego set?  How many in the smallest lego set? What is the median `val_star_rating`?

________________________________________________________________________________________________________________________________

## Getting Summary Statistics

Pandas also allows us to easily compute individual summary statistics using built-in methods.  Next, we'll get some practice using these methods. 

In the cell below, compute the median value of the `star_rating` column.

In [8]:
df.star_rating.median()

4.7

Next, get a count of the total number of values in `play_star_rating`.

In [9]:
df.play_star_rating.count()

10486

Now, compute the standard deviation of the `list_price` column.

In [10]:
df.list_price.std()

91.9804293059243

In [11]:
df.list_price.var()

8460.399375302139

In [12]:
df.list_price.var() ** .5

91.9804293059243

In [13]:
df.list_price.std() ** 2

8460.399375302139

If we bought every single lego set in this dataset, how many pieces would we have?  

> **Note**: If you truly want to answer this accurately, and are up for the challenge, try and remove duplicate lego-set entries before summing the pieces. That is, many of the lego sets are listed multiple times in the dataset above, depending on the country where it is being sold and other unique parameters. If you're stuck, just practice calculating the total number of pieces in the dataset for now.

In [21]:
df.loc[df.prod_id == 70920]

ages  list_price  num_reviews  piece_count  play_star_rating  \
190    7-14     29.9900         11.0        293.0               4.9   
776    7-14     29.9900         11.0        293.0               4.9   
955    7-14     37.9924         11.0        293.0               4.9   
1359   7-14     37.9924         11.0        293.0               4.9   
1525   7-14     42.6878         12.0        293.0               4.9   
1934   7-14     42.6878         12.0        293.0               4.9   
2098   7-14     54.8878         12.0        293.0               4.9   
2496   7-14     54.8878         12.0        293.0               4.9   
2718   7-14     27.2922         11.0        293.0               4.9   
3303   7-14     27.2922         11.0        293.0               4.9   
3481   7-14     40.6980         11.0        293.0               4.9   
3883   7-14     40.6980         11.0        293.0               4.9   
4049   7-14     47.9520         12.0        293.0               4.9   
4433   7-14     47.9520         12.0        293.0               4.9   
4598   7-14     48.7878         11.0        293.0               4.9   
4980   7-14     48.7878         11.0        293.0               4.9   
5145   7-14     41.8800         11.0        293.0               4.9   
5552   7-14     41.8800         11.0        293.0               4.9   
5722   7-14     48.7878         11.0        293.0               4.9   
6104   7-14     48.7878         11.0        293.0               4.9   
6270   7-14     54.8390         12.0        293.0               4.9   
6652   7-14     54.8390         12.0        293.0               4.9   
6820   7-14     42.6878         11.0        293.0               4.9   
7222   7-14     42.6878         11.0        293.0               4.9   
7388   7-14     41.9860         11.0        293.0               4.9   
7798   7-14     41.9860         11.0        293.0               4.9   
7961   7-14     45.1278         11.0        293.0               4.9   
8365   7-14     45.1278         11.0        293.0               4.9   
8533   7-14     42.6878         12.0        293.0               4.9   
8918   7-14     42.6878         12.0        293.0               4.9   
9081   7-14     54.8878         12.0        293.0               4.9   
9464   7-14     54.8878         12.0        293.0               4.9   
9630   7-14     58.3700         12.0        293.0               4.9   
10013  7-14     58.3700         12.0        293.0               4.9   
10180  7-14     54.8878         11.0        293.0               4.9   
10587  7-14     54.8878         11.0        293.0               4.9   
10757  7-14     42.5929         11.0        293.0               4.9   
11322  7-14     46.3971         11.0        293.0               4.9   
11686  7-14     46.3971         11.0        293.0               4.9   
11851  7-14     48.7878         12.0        293.0               4.9   
12232  7-14     48.7878         12.0        293.0               4.9   

                                   prod_desc  prod_id  \
190    Join Batman™ in a comical food fight!  70920.0   
776    Join Batman™ in a comical food fight!  70920.0   
955    Join Batman™ in a comical food fight!  70920.0   
1359   Join Batman™ in a comical food fight!  70920.0   
1525   Join Batman™ in a comical food fight!  70920.0   
1934   Join Batman™ in a comical food fight!  70920.0   
2098   Join Batman™ in a comical food fight!  70920.0   
2496   Join Batman™ in a comical food fight!  70920.0   
2718   Join Batman™ in a comical food fight!  70920.0   
3303   Join Batman™ in a comical food fight!  70920.0   
3481   Join Batman™ in a comical food fight!  70920.0   
3883   Join Batman™ in a comical food fight!  70920.0   
4049   Join Batman™ in a comical food fight!  70920.0   
4433   Join Batman™ in a comical food fight!  70920.0   
4598   Join Batman™ in a comical food fight!  70920.0   
4980   Join Batman™ in a comical food fight!  70920.0   
5145   Join Batman™ in a comical food fight!  70920.0  

In [4]:
udf = df.drop_duplicates(subset = ['prod_id'])

In [5]:
udf.loc[udf.prod_id == 70920]

ages  list_price  num_reviews  piece_count  play_star_rating  \
190  7-14       29.99         11.0        293.0               4.9   

                                 prod_desc  prod_id  \
190  Join Batman™ in a comical food fight!  70920.0   

                                        prod_long_desc review_difficulty  \
190  Help Batman™ fend off Egghead™ and Condiment K...           Average   

                     set_name  star_rating              theme_name  \
190  Egghead™ Mech Food Fight          4.7  THE LEGO® BATMAN MOVIE   

     val_star_rating country  
190              4.6      US

In [27]:
udf.piece_count.sum()

319071.0

Now, let's try getting the value for the 90% quantile.  Do this in the cell below.

In [28]:
udf.quantile(0.9)

list_price             99.99
num_reviews            35.00
piece_count           917.50
play_star_rating        5.00
prod_id             75822.70
star_rating             5.00
val_star_rating         5.00
Name: 0.9, dtype: float64

## Getting Summary Statistics on Categorical Data

For obvious reasons, most of the methods we've used so far only work with numerical data--there's no way to calculate the standard deviation of a column containing string values. However, there are some things that we can discover about columns containing categorical data. 

In the cell below, get the `.unique()` values contained within the `review_difficulty` column. 

In [6]:
df.review_difficulty.unique()

array(['Average', 'Easy', 'Challenging', 'Very Easy', nan,
       'Very Challenging'], dtype=object)

Now, let's get the `value_counts` for this column, to see how common each is. 

In [7]:
df.review_difficulty.value_counts()

Easy                4236
Average             3765
Very Easy           1139
Challenging         1058
Very Challenging       8
Name: review_difficulty, dtype: int64

As you can see, these provide us quick and easy ways to get information on columns containing categorical information.  


## Using `.applymap()`

When working with pandas DataFrames, we can quickly compute functions on the data contained by using the `applymap()` function and passing in a lambda function. 

For instance, we can use `applymap()` to return a version of the DataFrame where every value has been converted to a string.

In the cell below:

* Call our DataFrame's `.applymap()` function and pass in `lambda x: str(x)`
* Call our new `string_df` object's `.info()` method to confirm that everything has been cast to a string

In [8]:
string_df = df.applymap(lambda x: str(x))

In [9]:
string_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 14 columns):
ages                 12261 non-null object
list_price           12261 non-null object
num_reviews          12261 non-null object
piece_count          12261 non-null object
play_star_rating     12261 non-null object
prod_desc            12261 non-null object
prod_id              12261 non-null object
prod_long_desc       12261 non-null object
review_difficulty    12261 non-null object
set_name             12261 non-null object
star_rating          12261 non-null object
theme_name           12261 non-null object
val_star_rating      12261 non-null object
country              12261 non-null object
dtypes: object(14)
memory usage: 1.3+ MB


In [10]:
apply = df.apply(lambda x: str(x))

In [13]:
apply.head(2)

ages          0        6-12\n1        6-12\n2        6-12\n3...
list_price    0          29.99\n1          19.99\n2         ...
dtype: object

In [ ]:
df.piece_count.apply

In [ ]:
t

Note that everything--even the `NaN` values, has been cast to a string in the example above. 

Note that for pandas Series objects (such as a single column in a DataFrame), we can do the same thing using the `apply()` method.  

This is just one example of how we can quickly compute custom functions on our DataFrame--this will become especially useful when we learn how to **_normalize_** our datasets in a later section!

## Summary

In this lab, we learned how to:

* Understand and use the df.describe() and df.info() summary statistics methods
* Use built-in Pandas methods for calculating summary statistics (.mean(), .std(), .count(), .sum(), .mean(), .median(), .std(), .var() and .quantile())
* Apply a function to every element in a Series or DataFrame using s.apply() and df.applymap()